ddl/schema_test.go

```diff
@@ -226,14 +226,32 @@ func (s *testSchemaSuite) TestSchemaWaitJob(c *C) {
 	doDDLJobErr(c, schemaID, 0, model.ActionCreateSchema, []interface{}{dbInfo}, ctx, d2)
 }
 
-func testRunInterruptedJob(c *C, d *ddl, job *model.Job) {
+// runInterruptedJob should be called concurrently with restartWorkers
+func runInterruptedJob(c *C, d *ddl, job *model.Job, doneCh chan struct{}) {
 	ctx := mock.NewContext()
 	ctx.Store = d.store
 
-	done := make(chan error, 1)
-	go func() {
-		done <- d.doDDLJob(ctx, job)
-	}()
+	var (
+		history *model.Job
+		err     error
+	)
+
+	d.mu.RLock()
+	_ = d.doDDLJob(ctx, job)
```

Comment:
How about adding a check whether d is closed at every beginning of `doDDLJob`.
Avoid the case like that:
`doDDLJob` is called after`d.stop`, actually there is no worker running, so wait never ends in the loop.

Reply:
in the loop of `doDDLJob`, `case <-d.ctx.Done()` added in this PR will be exit branch. I'm afraid these lines will block if worker closed. Should we check `d.ctx.Done()` here?

https://github.com/pingcap/tidb/blob/547c5b14cf2162ca2a018135aca453e57222dc44/ddl/ddl.go#L466-L467

Reply:
yes, I think so, because of the d.ctx.Done() is triggered, the task won't be added into the queue successfully in `limitDDLJobs`.

 `err := <-task.err ` may hang here.

how check `d.ctx.Done()` at every beginning of `doDDLJob`?

Reply:
> how about checking `d.ctx.Done()` at every beginning of `doDDLJob`?

Maybe it's better to check at `err := <-task.err` and `d.limitJobCh <- task` to reduce gap between check and call.

Reply:
seems the same? ok